# Global Terrorism

Terrorism has been described variously as both a tactic and strategy; a crime and a holy duty; a justified reaction to oppression and an inexcusable abomination. Obviously, a lot depends on whose point of view is being represented. Terrorism has often been an effective tactic for the weaker side in a conflict. As an asymmetric form of conflict, it confers coercive power with many of the advantages of military force at a fraction of the cost. Due to the secretive nature and small size of terrorist organizations, they often offer opponents no clear organization to defend against or to deter.

This dataset contains 170,000 line items of the terrorism activities that happened between 1970 to 2016.With this extensive data,there is an opportunity for a data analyst to dissect and analyse the scenario and come out with interesting findings

## Import Libraries and Packages

In [1]:
import pandas as pd 
import seaborn as sns
import numpy as np
import os
import chart_studio
from chart_studio.plotly import iplot
import cufflinks as cf
import plotly.plotly as py
from plotly.graph_objs import *
import chart_studio.plotly
cf.set_config_file(offline=True, world_readable=True, theme='ggplot')
from subprocess import check_output
cf.go_offline()
cf.set_config_file(offline=False, world_readable=True)

#### Set Directory

In [2]:
os.getcwd() 
os.chdir(r"D:\OneDrive - Temple University\Python\Terroism") 

### Importing the Dataset

In [3]:
data = pd.read_csv("terrorism.csv", encoding='latin', low_memory=False) 
data.head(2)

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  1.970000e+11   1970       7     2        NaN         0        NaN       58   
1  1.970000e+11   1970       0     0        NaN         0        NaN      130   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  

[2 rows x 135 columns]

#### Brief view of the dataset
Initial dataset contain 135 columns, 181690 records and 3 data types used inn the dataset (float64(56), int64(21), object(58))

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Columns: 135 entries, eventid to related
dtypes: float64(56), int64(21), object(58)
memory usage: 187.1+ MB


Shape of dataset : 135 columns , 181619 rows

In [5]:
data.shape

(181691, 135)

Defining the frame so that we can easy see the data

In [6]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

#### Checking got Null Values 
In the initial dataset there are lot of null values

In [7]:
data.isnull().sum()

eventid                    0
iyear                      0
imonth                     0
iday                       0
approxdate            172452
extended                   0
resolution            179471
country                    0
country_txt                0
region                     0
region_txt                 0
provstate                421
city                     434
latitude                4556
longitude               4557
specificity                6
vicinity                   0
location              126196
summary                66129
crit1                      0
crit2                      0
crit3                      0
doubtterr                  1
alternative           152680
alternative_txt       152680
multiple                   1
success                    0
suicide                    0
attacktype1                0
attacktype1_txt            0
attacktype2           175377
attacktype2_txt       175377
attacktype3           181263
attacktype3_txt       181263
targtype1     

### Remaning the columns
Remainng columns in order to standardize gthe naming convention

In [8]:
data.rename(columns={'iyear':'Year','imonth':'Month','iday':'Day','country_txt':'Country',
                    'provstate':'state','natlty1_txt': 'Nationality','weaptype1_txt': 'WeaponType',
                     'region_txt':'Region','attacktype1_txt':'AttackType',
                    'target1':'Target','nkill':'Killed','nwound':'Wounded','summary':'Summary',
                    'gname':'Group','targtype1_txt':'Target_type','weaptype1_txt':'Weapon_type',
                    'motive':'Motive'},inplace=True)

Checking the datatype of all the selected columns

In [9]:
data.dtypes

eventid               float64
Year                    int64
Month                   int64
Day                     int64
approxdate             object
extended                int64
resolution             object
country                 int64
Country                object
region                  int64
Region                 object
state                  object
city                   object
latitude              float64
longitude             float64
specificity           float64
vicinity                int64
location               object
Summary                object
crit1                   int64
crit2                   int64
crit3                   int64
doubtterr             float64
alternative           float64
alternative_txt        object
multiple              float64
success                 int64
suicide                 int64
attacktype1             int64
AttackType             object
attacktype2           float64
attacktype2_txt        object
attacktype3           float64
attacktype

In [11]:
data=data[['Year','Month','Day','Country','state','Region','city',
                     'latitude','longitude','AttackType','Killed','Wounded', 'Nationality', 
                     'Weapon_type','Target','Summary','Group','Target_type',
                     'Motive']]

In [12]:
data.head(6)

Year  Month  Day             Country     state                       Region           city   latitude   longitude                      AttackType  Killed  Wounded         Nationality Weapon_type                     Target                                            Summary                               Group                  Target_type                                          Motive
0  1970      7    2  Dominican Republic       NaN  Central America & Caribbean  Santo Domingo  18.456792  -69.951164                   Assassination     1.0      0.0  Dominican Republic     Unknown               Julio Guzman                                                NaN                              MANO-D  Private Citizens & Property                                             NaN
1  1970      0    0              Mexico   Federal                North America    Mexico city  19.371887  -99.086624     Hostage Taking (Kidnapping)     0.0      0.0             Belgium     Unknown    Nadine Chaval, daughter                                                NaN  23rd of September Communist League      Government (Diplomatic)                                             NaN
2  1970      1    0         Philippines    Tarlac               Southeast Asia        Unknown  15.478598  120.599741                   Assassination     1.0      0.0       United States     Unknown                   Employee                                                NaN                             Unknown          Journalists & Media                                             NaN
3  1970      1    0              Greece    Attica               Western Europe         Athens  37.997490   23.762728               Bombing/Explosion     NaN      NaN       United States  Explosives               U.S. Embassy                                                NaN                             Unknown      Government (Diplomatic)                                             NaN
4  1970      1    0               Japan   Fukouka                    East Asia        Fukouka  33.580412  130.396361  Facility/Infrastructure Attack     NaN      NaN       United States  Incendiary             U.S. Consulate                                                NaN                             Unknown      Government (Diplomatic)                                             NaN
5  1970      1    1       United States  Illinois                North America          Cairo  37.005105  -89.176269                   Armed Assault     0.0      0.0       United States    Firearms  Cairo Police Headquarters  1/1/1970: Unknown African American assailants ...                  Black Nationalists                       Police  To protest the Cairo Illinois Police Deparment

In [13]:
data.isnull().sum()

Year                0
Month               0
Day                 0
Country             0
state             421
Region              0
city              434
latitude         4556
longitude        4557
AttackType          0
Killed          10313
Wounded         16311
Nationality      1559
Weapon_type         0
Target            636
Summary         66129
Group               0
Target_type         0
Motive         131130
dtype: int64

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 19 columns):
Year           181691 non-null int64
Month          181691 non-null int64
Day            181691 non-null int64
Country        181691 non-null object
state          181270 non-null object
Region         181691 non-null object
city           181257 non-null object
latitude       177135 non-null float64
longitude      177134 non-null float64
AttackType     181691 non-null object
Killed         171378 non-null float64
Wounded        165380 non-null float64
Nationality    180132 non-null object
Weapon_type    181691 non-null object
Target         181055 non-null object
Summary        115562 non-null object
Group          181691 non-null object
Target_type    181691 non-null object
Motive         50561 non-null object
dtypes: float64(4), int64(3), object(12)
memory usage: 26.3+ MB


#### Finding out the Counties majorly affected by the terrorist activities 
In Iraq, there are more than 24000 attackes have been registered which made it most affected country followed by Pakistan (14000+) and Afganistan (12500+)

In [15]:

data['Country'].value_counts()[:10][::-1].iplot(kind="barh",bargap=.5, title="", 
                                                colors="#182844", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top 10 countries faced terrorist attacks'))

#### Regionwise distribution
South Asia is the most effected region by terrorist activities followed by South America

In [33]:
data['Region'].value_counts()[1:10][::-1].iplot(kind="barh",bargap=.5, title="Top 10 cities faced terrorist attacks", 
                                                colors="mediumorchid", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top 10 Region faced terrorist attacks',
                                                                                margin=dict(l=200, r=50, t=80, b=50)))

#### Cities in which majorly attackes done
Baghdad is the most affatected one followed by Karachi. Baghdad suffered around 2.5 times more attacks compared to Karachi which is just next to it in the list

In [16]:
data['city'].value_counts()[1:10][::-1].iplot(kind="barh",bargap=.5, title="Top 10 cities faced terrorist attacks", 
                                              colors="#17becf", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top 10 cities faced terrorist attacks'))

#### Time Analysis
Year: 2014 was the year in which most of the attachs happened as compared to the last 15 years. and it graduaually decreases afterwards

Month: May months has relatively high attack frequency 

Day: Majority of the events happend in the middle of the month and decreased dransically at the end of the month

In [34]:
data.groupby('Year').size()[-15:].iplot(kind="line", title='Number of terrorist attacks in last 10 years', 
                                        colors="teal", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=500, height=500, 
                                                                                title='Year wise terrorist activity trends'))
data.groupby('Month').size()[1:].iplot(kind="line", 
                                        colors="teal", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=500, height=500, 
                                                                                title='Month wise terrorist attacks trends'))
data.groupby('Day').size()[1:].iplot(kind="line", 
                                        colors="teal", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=500, height=500, 
                                                                                title='Day wise terrorist attacks trends'))

#### Attacks Type

Majority of the attacks are done through arms followed by mass assasination 

In [43]:
data['AttackType'].value_counts()[1:10][::-1].iplot(kind="barh",bargap=.5, title="Top 10 attacktype faced terrorist attacks", 
                                                colors="teal", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Attack Type',
                                                                                margin=dict(l=300, r=50, t=80, b=50)))

#### Finding out the Groups reponsible for majority of the attacks 
Taliban has taken reponsibility for majority of the terrorsit attacks followed by ISIL


In [46]:
data['Group'].value_counts()[1:10][::-1].iplot(kind="barh",bargap=.5, 
                                                colors="#EEDFCC", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top 10 terrorist Goups',
                                                                                margin=dict(l=300, r=50, t=80, b=50)))

#### Target 

Civilians are the major targets of most of the terrorist activities

In [47]:
data['Target'].value_counts()[0:10][::-1].iplot(kind="barh",bargap=.5, 
                                                colors="#FFD39B", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top Targets of terrorist attacks'))

In [48]:
data['Nationality'].value_counts()[0:10][::-1].iplot(kind="barh",bargap=.5, 
                                                colors="#9BCD9B", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Nationality of terrorist attacks',
                                                                                margin=dict(l=200, r=50, t=80, b=50)))

#### Weapon Type Aanalysis
Explosive is the primary weapon used in attacks followed by firearms

In [49]:
data['Weapon_type'].value_counts()[0:10][::-1].iplot(kind="barh",bargap=.5, 
                                                colors="#CDC673", layout=Layout(paper_bgcolor='rgba(0,0,0,0)',
                                                                                plot_bgcolor='rgba(0,0,0,0)',
                                                                                width=800, height=500, 
                                                                                title='Top Weapon Type', 
                                                                                margin=dict(l=200, r=50, t=80, b=50)))

## Summary

1. Mojority of the attackes happend in South Asia (~45000)
2. Iraq is most affected country by attckes followed by Pakistan
3. Baghdad suffered around 2.5 times more attacks compared to Karachi which is just next to it in the list
4. Taliban is mostly responsible for terrorist activities followed by ISIL
5. Explosives are the main source of conducting attacks 

In [2]:
help(matplotlib)

NameError: name 'matplotlib' is not defined